hathor/p2p/sync_v2/manager.py

```diff
@@ -0,0 +1,1207 @@
+        """ Stop sync.
+        """
+        if not self._started:
+            raise Exception('NodeSyncBlock is already stopped')
+        self._started = False
+        self._lc_run.stop()
+
+    def get_cmd_dict(self) -> dict[ProtocolMessages, Callable[[str], None]]:
+        """ Return a dict of messages of the plugin.
+
+        For further information about each message, see the RFC.
+        Link: https://github.com/HathorNetwork/rfcs/blob/master/text/0025-p2p-sync-v2.md#p2p-sync-protocol-messages
+        """
+        return {
+            ProtocolMessages.GET_NEXT_BLOCKS: self.handle_get_next_blocks,
+            ProtocolMessages.GET_PREV_BLOCKS: self.handle_get_prev_blocks,
+            ProtocolMessages.BLOCKS: self.handle_blocks,
+            ProtocolMessages.BLOCKS_END: self.handle_blocks_end,
+            ProtocolMessages.GET_BEST_BLOCK: self.handle_get_best_block,
+            ProtocolMessages.BEST_BLOCK: self.handle_best_block,
+            ProtocolMessages.GET_BLOCK_TXS: self.handle_get_block_txs,
+            ProtocolMessages.GET_TRANSACTIONS_BFS: self.handle_get_transactions_bfs,
+            ProtocolMessages.TRANSACTION: self.handle_transaction,
+            ProtocolMessages.TRANSACTIONS_END: self.handle_transactions_end,
+            ProtocolMessages.GET_PEER_BLOCK_HASHES: self.handle_get_peer_block_hashes,
+            ProtocolMessages.PEER_BLOCK_HASHES: self.handle_peer_block_hashes,
+            ProtocolMessages.STOP_BLOCK_STREAMING: self.handle_stop_block_streaming,
+            ProtocolMessages.GET_TIPS: self.handle_get_tips,
+            ProtocolMessages.TIPS: self.handle_tips,
+            ProtocolMessages.TIPS_END: self.handle_tips_end,
+            # XXX: overriding ReadyState.handle_error
+            ProtocolMessages.ERROR: self.handle_error,
+            ProtocolMessages.GET_DATA: self.handle_get_data,
+            ProtocolMessages.DATA: self.handle_data,
+            ProtocolMessages.RELAY: self.handle_relay,
+            ProtocolMessages.NOT_FOUND: self.handle_not_found,
+        }
+
+    def handle_not_found(self, payload: str) -> None:
+        """ Handle a received NOT-FOUND message.
+        """
+        # XXX: NOT_FOUND is a valid message, but we shouldn't ever receive it unless the other peer is running with a
+        #                modified code or if there is a bug
+        self.log.warn('not found? close connection', payload=payload)
+        self.protocol.send_error_and_close_connection('Unexpected NOT_FOUND')
+
+    def handle_error(self, payload: str) -> None:
+        """ Override protocols original handle_error so we can recover a sync in progress.
+        """
+        assert self.protocol.connections is not None
+        # forward message to overloaded handle_error:
+        self.protocol.handle_error(payload)
+
+    def update_synced(self, synced: bool) -> None:
+        self._synced = synced
+
+    @inlineCallbacks
+    def run_sync(self) -> Generator[Any, Any, None]:
```

Comment:
That's the way to implement async using twisted. Not sure we can do better than this.

Reply:
I did some digging on this (yes, I may be a type freak, but I do think typing is _that_ important), and here's what I found out.

The best we can easily do right now is just typing the variables we assign the yielded deferreds to. In this `run_sync()` method we don't have this, but in `find_best_common_block()` for example, we have:

```python
block_height_list = yield self.get_peer_block_hashes(heights)
```

So in my opinion, we should type hint the variable there to `block_height_list: list[tuple[int, bytes]]`, which is the type returned from `get_peer_block_hashes()`'s deferred. Type checking passes and we get guarantees for using `block_height_list` downwards. I think we should do this explicit typing every time we use async Twisted with `@inlineCallbacks` and `yield`.

Other than that, we can change the method return type hints from `Generator[Any, Any, None]` to `Generator[Deferred[Any], Any, None]`. But that doesn't help much... The `Generator` [type parameters are](https://docs.python.org/3.9/library/typing.html#typing.Generator):

1. `YieldType`: the type we yield, that is, in the context of async Twisted, always a `Deferred[Something]`
2. `SendType`: the type the yield sends _back_ to us, that is, the `Something` that the `Deferred` will eventually return
3. `ReturnType`: what the function returns, I think we'll never use this with async Twisted so it's always `None`

So in theory we _could_ type hint the returned `Generator` correctly, in a way that the `yield` types would be automatically inferred and we wouldn't have to explicitly type the variables as I suggested above.

For `find_best_common_block()`, then, the return type would be:

```python
```

Reply:
I did some digging on this (yes, I may be a type freak, but I do think typing is _that_ important), and here's what I found out.

The best we can easily do right now is just typing the variables we assign the yielded deferreds to. In this `run_sync()` method we don't have this, but in `find_best_common_block()` for example, we have:

```python
block_height_list = yield self.get_peer_block_hashes(heights)
```

So in my opinion, we should type hint the variable there to `block_height_list: list[tuple[int, bytes]]`, which is the type returned from `get_peer_block_hashes()`'s deferred. Type checking passes and we get guarantees for using `block_height_list` downwards. I think we should do this explicit typing every time we use async Twisted with `@inlineCallbacks` and `yield`.

Other than that, we can change the method return type hints from `Generator[Any, Any, None]` to `Generator[Deferred[Any], Any, None]`. But that doesn't help much... The `Generator` [type parameters are](https://docs.python.org/3.9/library/typing.html#typing.Generator):

1. `YieldType`: the type we yield, that is, in the context of async Twisted, always a `Deferred[Something]`
2. `SendType`: the type the yield sends _back_ to us, that is, the `Something` that the `Deferred` will eventually return
3. `ReturnType`: what the function returns, I think we'll never use this with async Twisted so it's always `None`

So in theory we _could_ type hint the returned `Generator` correctly, in a way that the `yield` types would be automatically inferred and we wouldn't have to explicitly type the variables as I suggested above.

For `find_best_common_block()`, then, the return type would be:

```python
Generator[list[tuple[int, bytes]], list[tuple[int, bytes]], None]
```

Of course that's pretty convoluted, but we could create some type alias for it, maybe something like `AsyncTwisted[T]` and then use it as `AsyncTwisted[list[tuple[int, bytes]]]` in this case.

Unfortunately, there are two problems with this:

1. Mypy complains if we do that, because Twisted's `@inlineCallbacks` has bad typing, by using `object` instead of generics. That's easily fixed by creating our own wrapper around `@inlineCallbacks` that does nothing, just fixes its typing with `cast`. But it's starting to get a bit overkill...
2. The real problem: if we use `yield` multiple times in a function, all their different yielded types must be in the `Generator` arguments, using `Union`. That defeats the whole purpose of typing as we can't know which `yield` is yielding which type. I could not find a way around this 😢

So, summing up my conclusions:

1. We should explicitly type hint all variables assigned to async yields
2. We should change async return types from `Generator[Any, Any, None]` to `Generator[Deferred[Any], Any, None]` (and we could type alias that)
3. We _could_ use completely typed `Generator`s for functions that only use one `yield`
4. There's no better way to type `Generator`s for functions that use multiple `yield`s